pkg/apis/projectcalico/v3/nodestatus.go

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2021 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v3
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	KindCalicoNodeStatus     = "CalicoNodeStatus"
+	KindCalicoNodeStatusList = "CalicoNodeStatusList"
+)
+
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CalicoNodeStatusList is a list of CalicoNodeStatus resources.
+type CalicoNodeStatusList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Items []CalicoNodeStatus `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type CalicoNodeStatus struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Spec   CalicoNodeStatusSpec   `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+	Status CalicoNodeStatusStatus `json:"status,omitempty" protobuf:"bytes,2,opt,name=status"`
+}
+
+// CalicoNodeStatusSpec contains the specification for a CalicoNodeStatus resource.
+type CalicoNodeStatusSpec struct {
+	// The node name identifies the Calico node instance for node status.
+	Node string `json:"node,omitempty" validate:"required,name"`
+
+	// Classes specifies type of information CalicoNodeStatus should contain.
+	Classes []NodeStatusClassType `json:"classes,omitempty"`
+
+	// UpdateIntervalInSeconds is the period at which CalicoNodeStatus should be updated.
+	// Set to 0 to disable CalicoNodeStatus refresh. [Default: 10]
+	UpdateIntervalInSeconds int `json:"updateIntervalInSeconds,omitempty"`
+}
+
+// CalicoNodeStatusStatus defines the observed state of CalicoNodeStatus.
+// No validation needed for status since it is updated by Calico.
+type CalicoNodeStatusStatus struct {
+	// UpdateTimestamp is a timestamp representing the server time when CalicoNodeStatus object
+	// last updated. It is represented in RFC3339 form and is in UTC.
+	UpdateTimestamp metav1.Time `json:"updateTimestamp,omitempty"`
+
+	// AdditionalInfo is a a human-readable description of the status of last update.
+	AdditionalInfo string `json:"additionalInfo,omitempty"`
```

Comment:
Should this be a map[string]string so that we can add arbitrary metadata to it?

Remind me what sort of information we're planning to put here today?

Reply:
This is just an additional information about the status of last update. It could contain error message if the last update failed for any reason.

Reply:
Kubernetes generally uses `Conditions` in status to report on things like this - we should follow that pattern instead of adding an opaque string. It's a more structured way of reporting messages intended for the end-user / a human.